src/main/java/reactor/core/publisher/MonoFilterWhen.java

```diff
@@ -45,7 +51,269 @@
+
+		static final int STATE_FRESH         = 0; //the source value is fresh
+		static final int STATE_RUNNING       = 1; //the filter is running asynchronously
+		static final int STATE_RESULT        = 2; //the output Mono is resolved
+
+		MonoFilterWhenSubscriber(Subscriber<? super T> actual, Function<? super T, ? extends Publisher<Boolean>> asyncPredicate) {
+			super(actual);
+			this.asyncPredicate = asyncPredicate;
+		}
+
+		@Override
+		public void onSubscribe(Subscription s) {
+			if (Operators.validate(upstream, s)) {
+				upstream = s;
+				actual.onSubscribe(this);
+				s.request(1);
+			}
+		}
+
+		@Override
+		public void onNext(T t) {
+			if (filterState == STATE_FRESH) {
+				setValue(t);
+				Publisher<Boolean> p;
+
+				try {
+					p = Objects.requireNonNull(asyncPredicate.apply(t),
+							"The asyncPredicate returned a null value");
+				}
+				catch (Throwable ex) {
+					Exceptions.throwIfFatal(ex);
+						//TODO not really needed as we don't expect to get there if the Mono source errors
+					Exceptions.addThrowable(ERROR, this, ex);
+					filterState = STATE_RESULT;
+					super.onError(error);
+					return;
+				}
+
+				if (p instanceof Callable) {
+					Boolean u;
+
+					try {
+						u = ((Callable<Boolean>) p).call();
+					}
+					catch (Throwable ex) {
+						Exceptions.throwIfFatal(ex);
+						//TODO not really needed as we don't expect to get there if the Mono source errors
+						Exceptions.addThrowable(ERROR, this, ex);
+						filterState = STATE_RESULT;
+						super.onError(error);
+						return;
+					}
+
+					if (u != null && u) {
+						actual.onNext(t);
+					}
+				}
+				else {
+					FilterWhenInner inner = new FilterWhenInner(this, !(p instanceof Mono));
+					if (ASYNC_FILTER.compareAndSet(this, null, inner)) {
+						filterState = STATE_RUNNING;
+						p.subscribe(inner);
+					}
+				}
+			}
+			//TODO drop?
+		}
+
+		@Override
+		public void onComplete() {
+			if (filterState == STATE_FRESH) {
+				//there was no value, we can complete empty
+				super.onComplete();
+			}
+			//otherwise just wait for the inner filter to apply, rather than complete too soon
+		}
+
+		@Override
+		public void onError(Throwable t) {
+			Exceptions.addThrowable(ERROR, this, t);
+			if (filterState != STATE_RUNNING) {
+				super.onError(t);
+			}
+		}
+
+		@Override
+		public Throwable getError() {
+			return error;
+		}
+
+		@Override
+		public void cancel() {
+			if (super.state != CANCELLED) {
+				super.cancel();
+				upstream.cancel();
+				cancelInner();
+			}
+		}
+
+		void cancelInner() {
+			FilterWhenInner a = asyncFilter;
+			if (a != INNER_CANCELLED) {
+				a = ASYNC_FILTER.getAndSet(this, INNER_CANCELLED);
+				if (a != null && a != INNER_CANCELLED) {
+					a.cancel();
+				}
+			}
+		}
+
+		void innerResult(Boolean item) {
+			filterState = STATE_RESULT;
+			if (item != null && item) {
+				//will reset the value with itself, but using parent's `value` saves a field
+				complete(value);
```

Comment:
`complete(T)` calls `setValue(t)`, which we already use back in `onNext`. This is idempotent though, and avoids the introduction of a second value field just to temporarily store the source value while the async filter resolves.